[Feature] Add support for other bip39 language wordlists

abhigit-saha opened the pull request:
## Description

Added support for BIP39 wordlists in Spanish, Portuguese, Italian, and French, as discussed in the issue. Resolves #679.

Incorporation of other wordlists would help non-English speakers to load seeds from other wallets that they have created which don't have English wordlists, well as generate seed words that are easier to remember for them. 

Screenshots: 

- Wordlist Language Selection
![Screenshot from 2025-04-07 17-15-34](https://github.com/user-attachments/assets/ab20aea6-7833-43cc-9a7e-f5bf95c193bf)

- Non-English wordlist warning
![Screenshot from 2025-04-07 17-15-54](https://github.com/user-attachments/assets/f022f587-01cf-4d55-a660-fa798eb6eb57)

- Seed Word Entry Screen with French wordlist
![Screenshot from 2025-04-07 17-16-10](https://github.com/user-attachments/assets/8681fc78-7986-4a5f-8170-b01022a2086d)

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before submitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [x] Yes
- [ ] N/A

I have tested this PR on the following platforms/os:

- [ ] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [x] Other

fedebuyito commented:
Hi, nice job! While I was testing it I found differents seeds for same words combination:

abandon x 11 + about (ENGLISH):
![image](https://github.com/user-attachments/assets/d370093f-cc07-48df-a3bb-d067d3dab8db)

ábaco x 11 + abierto (SPANISH):
![image](https://github.com/user-attachments/assets/b6a63075-72c1-4473-9f90-cd80fb149378)

PD: tested on python 3.12 and fine!
![image](https://github.com/user-attachments/assets/ac5238fd-8132-43c4-8005-3af65c6e16aa)

abhigit-saha commented:
Hi @fedebuyito , I think those two should yield the same _entropy_ , but since the seed depends upon the normalized seed words themselves, they must be different. Do let me know if I have mixed up something

polespinasa commented on src/seedsigner/controller.py:
```diff
@@ -148,7 +148,7 @@ def get_instance(cls):
             return cls.configure_instance()
     
-
+    
```
Don't add empty whitespaces

polespinasa commented on src/seedsigner/controller.py:
```diff
@@ -239,7 +239,7 @@ def pop_prev_from_back_stack(self):
         self.back_stack = BackStack()
 
-
+    
```
Empty line - Same

polespinasa commented on src/seedsigner/gui/keyboard.py:
```diff
@@ -302,7 +302,7 @@ def render_keys(self, selected_letter=None):
             that section, as in when changing `active_keys` or swapping to alternate
             charsets (e.g. alpha to special symbols).
-
+            
```
Empty line - Same

polespinasa commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -130,9 +135,12 @@ def calc_possible_alphabet(self, new_letter = False):
+                word = self.remove_accents(word);
                 if len(word)-1 >= letter_num:
-                    possible_letters.append(word[letter_num])
+                    # possible_letter = unicodedata.normalize('NFKD', word[letter_num]);
```
nit: don't let commented code

polespinasa commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -151,7 +161,7 @@ def render_possible_matches(self, highlight_word=None):
         # Render the possible matches to a temp ImageDraw surface and paste it in
         # BUT render the currently highlighted match as a normal Button element
-
+        
```
Same - New line

polespinasa commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -340,7 +350,6 @@ def _run(self):
                     self.text_entry_display.cur_text = ''.join(self.letters)
                     self.text_entry_display.render()
                     self.renderer.show_image()
-
```
Same - Remove line

polespinasa commented on src/seedsigner/models/seed.py:
```diff
@@ -5,6 +5,8 @@
 
 from binascii import hexlify
 from embit import bip39, bip32, bip85
+
```
remove empty line

polespinasa commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -391,7 +405,7 @@ class SettingsDefinition:
                       display_name=_mft("Mnemonic language"),
                       type=SettingsConstants.TYPE__SELECT_1,
-                      visibility=SettingsConstants.VISIBILITY__HIDDEN,
+                    #   visibility=SettingsConstants.VISIBILITY__HIDDEN,
```
don't let commented code

polespinasa commented on src/seedsigner/views/seed_views.py:
```diff
@@ -164,7 +164,7 @@ class LoadSeedView(View):
     TYPE_ELECTRUM = ButtonOption("Enter Electrum seed", FontAwesomeIconConstants.KEYBOARD)
     CREATE = ButtonOption("Create a seed", SeedSignerIconConstants.PLUS)
-
+    
```
same

polespinasa commented on src/seedsigner/views/seed_views.py:
```diff
@@ -183,7 +184,7 @@ def run(self):
             button_data=button_data
         )
-
+        
```
same

polespinasa commented on src/seedsigner/views/tools_views.py:
```diff
@@ -220,7 +220,7 @@ def __init__(self, total_rolls: int):
         self.total_rolls = total_rolls
     
-
+    
```
same

polespinasa commented on src/seedsigner/views/tools_views.py:
```diff
@@ -346,7 +348,7 @@ def __init__(self, coin_flips: str = None):
         #   * 7 bits to a 12-word seed (plus 4-bit checksum)
         from seedsigner.helpers import mnemonic_generation
-
+        
```
same

polespinasa commented on src/seedsigner/views/view.py:
```diff
@@ -122,14 +122,14 @@ def run(self, **kwargs) -> 'Destination':
     clear_history: bool = False         # Optionally clears the back_stack to prevent "back"
 
-
+    
```
same

polespinasa commented on src/seedsigner/views/view.py:
```diff
@@ -122,14 +122,14 @@ def run(self, **kwargs) -> 'Destination':
     """
         Basic struct to pass back to the Controller to tell it which View the user should
-        be presented with next.
+        be presented with next.         
```
remove empty spaces

polespinasa commented on tests/test_seed.py:
```diff
@@ -18,6 +20,7 @@ def test_seed():
 	# TODO: Not yet supported in new implementation
 	# seed.set_wordlist_language_code("es")
 	
+
```
same

polespinasa reviewed (commented):
Thanks for working on this.
Didn't try but lgtm, will test later.

One general comment. Please don't add or remove empty lines, it makes more difficult to review. Keep changes to the PR scope.
Also, you introduced some whitespaces. Try to follow good lint practices.

abhigit-saha commented:
> Thanks for working on this.
> Didn't try but lgtm, will test later.
> 
> One general comment. Please don't add or remove empty lines, it makes more difficult to review. Keep changes to the PR scope.
> Also, you introduced some whitespaces. Try to follow good lint practices.

Apologies for the slightly messy formatting, I'll be sure to fix that

polespinasa commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -122,6 +121,12 @@ def __post_init__(self):
+        import unicodedata
+        return ''.join(
+            c for c in unicodedata.normalize('NFKD', text) 
+            if unicodedata.category(c) != 'Mn'
```
I think this can be `if not unicodedata.combining(c)`

abhigit-saha replied:
I think both should work

polespinasa replied:
I agree but maybe looks cleaner. Feel free to ignore it if you prefer to keep it as it is.

kdmukai replied:
Comparison to a string literal definitely has to be avoided here. 

Also these lines need explanatory comments. Assume other devs (me!!) have no idea what `NFKD` means nor why it's necessary to normalize to it.

And I definitely have no idea what it means for the category to be `Mn`.

Obviously you can't explain everything about `unicodedata` in a short comment (at some point other devs just need to google stuff to learn more), but see other examples all over the code where less common python operations are briefly explained.

polespinasa reviewed (commented):
Generate seed from camera is not using the wordlist specified on the config:
![image](https://github.com/user-attachments/assets/8832d8d5-fda3-481a-8d7a-a89eb556c842)

Using dices works.

Bicaru20 commented:
Hello, nice job! 
I had a couple of errors though:
Just as @polespinasa mentioned, I also got an error when creating the seed from the camera:
![Screenshot from 2025-04-07 19-33-38](https://github.com/user-attachments/assets/f918fda4-2a96-4fee-83b5-d0c894631f57)
When using the `Calc 12th word ` it also looks as if it is not using the wordlist specified:
![Screenshot from 2025-04-07 19-24-29](https://github.com/user-attachments/assets/92017756-625a-498f-8aa9-34660cb63cad)
With the dice rolls I had no problem:
![Screenshot from 2025-04-07 19-40-45](https://github.com/user-attachments/assets/9119ec88-5ac0-4e4a-bcce-2faef6de78d9)
I have done the tests in the emulator, with python 3.12

Bicaru20 commented:
Regarding the error when creating the seed from the camera, I think I might have found the error:
In` tools_view.py` at line 164, when getting the `mnemonic` form the hash of the image:
```python
 mnemonic = mnemonic_generation.generate_mnemonic_from_bytes(final_hash)
``` 
You should also be passing the language world list that the user has chosen:
```python
        mnemonic = mnemonic_generation.generate_mnemonic_from_bytes(final_hash, self.settings.get_value(SettingsConstants.SETTING__WORDLIST_LANGUAGE))

``` 
Without this, the words you are generating are in English by default, causing the error.
With this change, I didn't get the error when creating the seed from the camera.
I hope this helps!

kdmukai commented:
Pretty great review feedback, testing, AND debugging help so far!!!

Awesome to see more people putting more eyes on PRs and the project in general!!

fedebuyito commented:
> Hi @fedebuyito , I think those two should yield the same _entropy_ , but since the seed depends upon the normalized seed words themselves, they must be different. Do let me know if I have mixed up something

As we discussed on telegram you are right, @abhigit-saha . Thank you for the clarification. I had a misconception about that.

I add test for python 3.10, as well:
![image](https://github.com/user-attachments/assets/24c49469-bcd0-41c0-ad09-ae8fa391e6ae)

Regarding the seeds generated it seems to be ok. I tested against Sparrow Wallet.

kdmukai commented on src/seedsigner/helpers/bip39/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .spanish_list import WORDLIST__SPANISH
+from .french_list import WORDLIST__FRENCH
+from .italian_list import WORDLIST__ITALIAN
+from .portuguese_list import WORDLIST__PORTUGUESE
```
It feels a bit bloat-y to import all the lists when in most cases users will need only ONE of them.

I haven't looked through the whole PR so maybe this is already mostly included, but I think it might be better to have some "get_wordlist()" utility that only does the `import` within an `if`/`elif`.

I don't think importing all of them would have any impact on load times or RAM usage since SeedSigner OS loads everything into RAM at boot time anyway. But for the eventual microcontroller port (e.g. esp32-s3, rp2350), we definitely need to consider the much more limited RAM environment (likely 8MB).

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -110,10 +109,11 @@ def set_passphrase(self, passphrase: str, regenerate_seed: bool = True):
+    
 
     def set_wordlist_language_code(self, language_code: str):
         # TODO: Support other BIP-39 wordlist languages!
```
Remove the `#TODO` if the thing TO DO is done!

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -193,20 +194,44 @@ def run(self):
         elif button_data[selected_menu_num] == self.TYPE_12WORD:
             self.controller.storage.init_pending_mnemonic(num_words=12)
-            return Destination(SeedMnemonicEntryView)
+            if wordlist_language_code=="en": 
```
Always use constants instead of literal strings.

kdmukai replied:
Also note typical style usage around `==`: space before and after in conditionals.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -193,20 +194,44 @@ def run(self):
+        button_data = [self.CONTINUE, self.BACK]
+        ret = self.run_screen(
+            seed_screens.SeedWordsLanguageWarningScreen, 
+            title=_("Non-English wordlist warning"),
```
My fault for not finishing our UX style guide, but titles should be in "title case", meaning that every word's first letter is capitalized.

Buttons (see your `BACK` button) should be in "sentence case" (only first letter of first word is capitalized).

Our UX definitely is not yet standardized to these guidelines, but we will make an effort to do so soon-ish.

kdmukai commented on tests/test_seed.py:
```diff
@@ -27,15 +30,15 @@ def test_seed():
-	# seed.set_passphrase("test")
+	mnemonic = "hebilla cría truco tigre gris llenar folio negocio laico casa tieso eludir".split()
+
+	seed_es = Seed(mnemonic=mnemonic, wordlist_language_code="es", passphrase="test")
```
Constant vs literal string "es".

kdmukai commented:
@easyuxd I can't recall: do we ever say "Go back" or do we let the back button suffice?

(see screenshot in PR description at top)

easyuxd commented:
@kdmukai For these warning screens, we only provide a Continue button and rely on the Back arrow for back navigation. So it's best to remove that extra button.

(We do have some conflicting language with "Continue" vs "I understand" -- I think Continue is clearer in these scenarios.)

![warning-screens](https://github.com/user-attachments/assets/6b39d68f-1398-4f8c-bc08-42e02719a2c9)

abhigit-saha commented:
Hi all, thank you for the feedback. I have considered all of the reviews and comments and hopefully all of the changes are there in place now...

I have changed the warning screen to be closer to the existing warning screens as well, 
![image](https://github.com/user-attachments/assets/24f198e3-aa2b-42c4-ac07-89bb70081968)

polespinasa commented:
Great! Will re-review.

Let me give you one tip. There's no need to keep different commits for each change. Normally 1 commit per functionality is great and facilitates the review as all changes are in the same place. You avoid seeing all the unnecessary changes (like the adding empty spaces and then removing them again). 

So you could squash all commits into the first one so everything will look cleaner and will be easier to review.

abhigit-saha commented:
Clarification: The recent commits have led to the removal of whitespace that was previously committed as well, due to my vscode settings configuration. It was difficult for me to undo the whitespace changes I did and not overwrite the already existing ones, thus the diff shows many such changes.

Bicaru20 reviewed (commented):
When testing the new changes, I've got the same error on the `Calc 12th/24th word ` option. I've been debuting, and I think I found the problems. I led you a comment on the code of what I think you should change. 
Also, you should also change the line 55 of the file `mnemonic_generation.py`, when you convert back the bytes into mnemonic. As it is right now:
`return bip39.mnemonic_from_bytes(mnemonic_bytes).split()`
you are telling the function to return to mnemonic in the default language (english). You must specify the language to avoid any errors:
` return bip39.mnemonic_from_bytes(mnemonic_bytes, wordlist= Seed.get_wordlist(wordlist_language_code)).split()`
These changes worked for me in the emulator.

The seed generator with the camera works fine.

I hope this helps!!

Bicaru20 commented on src/seedsigner/models/seed_storage.py:
```diff
@@ -79,26 +79,26 @@ def get_pending_mnemonic_word(self, index: int) -> str:
                 seed = ElectrumSeed(self._pending_mnemonic)
             else:
-                seed = Seed(self._pending_mnemonic)
+                seed = Seed(self._pending_mnemonic, wordlist_language_code)
```
Here, you should specify what argument is the variable `wordlist_language_code` referring to. As it is right now, is referring to the variable `passphrase`, and I believe it should refer to the variable `wordlist_language_code`. It should be like this:
`seed = Seed(self._pending_mnemonic, wordlist_language_code= wordlist_language_code)`

abhigit-saha commented:
I had a cruel time finally figuring out my workflow, with a lot of time being spent on trying to make my local dev-tool repo and the fork have the same changes. Currently it is not perfect, but should work. Earlier commits were leading to the same errors because of it (despite having fixed these issues then itself). I have made the required changes and cherry-picking the commits into my local dev-tool branch did fix all the issues that others were facing. Hopefully it does work for others as well, thank you all for the quick feedback

fedebuyito commented:
> I had a cruel time finally figuring out my workflow, with a lot of time being spent on trying to make my local dev-tool repo and the fork have the same changes. Currently it is not perfect, but should work. Earlier commits were leading to the same errors because of it (despite having fixed these issues then itself). I have made the required changes and cherry-picking the commits into my local dev-tool branch did fix all the issues that others were facing. Hopefully it does work for others as well, thank you all for the quick feedback

Nice! I hope the script can to help for everyone. As it is on "beta" testing (waiting to be used/adapted on more distros and OS) and would miss some improvements I have not do the PR on ss main repo yet wich would be more easy to pull and use for devs, I think. I am available for any doubts or suggests you have.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -223,6 +254,8 @@ def run(self):
             initial_letters=list(self.cur_word) if self.cur_word else ["a"],
             wordlist=Seed.get_wordlist(wordlist_language_code=self.settings.get_value(SettingsConstants.SETTING__WORDLIST_LANGUAGE)),
+            #TODO: add a standardized way to get the possible alphabet
+            possible_alphabet=_("abcdefghijklmnopqrstuvwxyz"),
```
This should NOT be wrapped for translation.

If a non-english wordlist is selected, we need to be explicit in the code what the expected character set is. Wrapping it for translation means that translators in EVERY language would feel compelled to provide their language's alphabet, even though the VAST MAJORITY of languages do not have their own bip39 wordlists. The Bulgarian translators would provide a Bulgarian alphabet, but that would break our bip39 implementation since there's no Bulgarian wordlist.

Even worse, if a Korean user needed to restore an English bip39 mnemonic, this translation wrapping would return the alphabet provided by the Korean translators; users would have to switch the entire UI to English in order to enter their mnemonic.

Also, per the TODO comment, I would have the `possible_alphabet` defined for each language in a constant either in `SettingsConstants` (hmm, it's not really a Setting though) or in `Seed` where the wordlist lookup is provided.

That being said, I don't love that the wordlist lookup is in `Seed`. Perhaps a `util` module within `helpers.bip39`?

kdmukai commented:
The whitespace changes look reasonable to me (though I have not carefully reviewed all of them), BUT this PR would be MUCH better if none of them were included.

Ideally a PR would be the minimum set of changes required to implement a new feature or fix a bug. Of course I have OFTEN violated this principle in horrible ways, unfortunately. But especially for PRs from new contributors, the simpler the PR, the better.

(also please note that I tend to just state any critiques without necessarily adding extra encouragement, especially right now when I'm very time/brain constrained. If I thought someone's work was not worthy, I would just "NACK" it and move on. Providing comments means that I want to see this keep progressing so it can make it to the finish line)

kdmukai commented on src/seedsigner/helpers/bip39/__init__.py:
```diff
@@ -0,0 +1,24 @@
+            from .italian_list import WORDLIST__ITALIAN
+            return WORDLIST__ITALIAN
+        case SettingsConstants.WORDLIST_LANGUAGE__PORTUGUESE:
+            from .portuguese_list import WORDLIST__PORTUGUESE
```
I think this file naming convention might have originally been my idea, but now that I look at it again, I think it would be better to use the language code names.

"en" instead of "english".

And maybe hide them further in a `helpers.bip39.wordlists` dir (especially if we do add a `helpers.bip39.util` or something similar; I like hiding away things that I'll never have to look at when there are things I might have to dig up).

Also the English wordlist is part of `embit` so I would not duplicate it here.

polespinasa replied:
> Also the English wordlist is part of `embit` so I would not duplicate it here.

IMHO, I would let it duplicated. I think it's better to have all lists in the same place. I would agree to just add all wordlists to `embit` but I don't think that would be feasible. I don't think having this duplicated is that bad.

kdmukai replied:
My thinking is more about avoiding any whiff of anything possibly being proprietary or out of step with the standards.

The English bip39 wordlist is unspeakably important. I prefer that the list be at least one step removed from our direct edit / merge control.

That being said, a basic sanity check that we're not getting a malicious wordlist from `embit` would also be a good idea.

kdmukai commented:
Another thought that I'm not sure how to feel about: Should the wordlists go in the `seedsigner-translations` repo instead?

https://github.com/SeedSigner/seedsigner-translations

It seems likely that eventually we'll put the translations in their own partition when we create a SeedSigner OS image. The hope is to minimize the RAM load that delays initial startup. The unnecessary bloat of these different wordlists seems to be a perfect fit for the same treatment. Plus we'll already have specific subdirs there for supported language codes.

fedebuyito commented:
Hi Keith, I agree. Here are catalan wordlist, too (asked initially in issue). 

https://github.com/polespinasa/seedsigner/blob/bip39-multilang/src/seedsigner/resources/bip39Wordlist/catalan_wordlist.py

abhigit-saha committed:
Add BIP39 wordlist support for other languages

abhigit-saha committed:
Merge 'origin/dev' into bip39wordlists

abhigit-saha committed:
Changed language variable names to concise language code names

abhigit-saha committed:
Added missing arguments to SeedWordlistLanguageWarningView

abhigit-saha committed:
Removed wordlist lookup from Seed class, replaced with a helper method

abhigit-saha committed:
Moved wordlists to a separate module, moved bip39 functions to util.py

abhigit-saha committed:
Replaced hardcoded charset using get_possible_alphabet

abhigit-saha committed:
Seed phrase verification now displays words of the same language as wordlist

abhigit-saha committed:
Hid the mnemonic options in Advanced Settings and added Warnings

abhigit-saha committed:
Added flow based tests

abhigit-saha committed:
Merge branch 'SeedSigner:dev' into bip39wordlists

abhigit-saha committed:
Minor cleanups

abhigit-saha committed:
Minor Cleanup

abhigit-saha committed:
Merge branch 'SeedSigner:dev' into bip39wordlists

abhigit-saha committed:
Replaced english wordlist with embit's and added sanity checks

abhigit-saha committed:
Merge branch 'bip39wordlists' of https://github.com/abhigit-saha/seedsigner into bip39wordlists

abhigit-saha committed:
Removed English Wordlist

abhigit-saha commented:
The following are the changes that have been recently made: 
1.  Changed the language variable names to their codes (as per keith's suggestion) 
2. Added a helper method that gives the required wordlist after passing the ```wordlist_language_code``` parameter. Only loads the wordlist for a language when that specific language is called. 
3. Hid the wordlists in a separate file inside ```/helpers/bip39 ```named ``wordlists```
4.  Replaced the hardcoded charset with a function to get the charset given the wordlist language. 
5. Edited the seed phrase verification screen to show the wordlists of the currentlly selected mnemonic language. 
6. As per Keith, hid the mnemonic language option inside advanced settings with appropriate warnings (would be making the warnings a bit more concise) 
7.  Added flow based tests where the non-english wordlists would affect using ```parametrize``` decorator available from ```pytest```
8. Added a basic sanity check for the English wordlist that we are getting through embit 

(note: i'd be squashing the commits together but right now it's just for ease of viewing the changes)

abhigit-saha commented:
However, still while going to a previously entered word in the seedphrase entry flow, (eg. from 4th to the 3rd) the candidate words that appear at the right are not being rendered properly in that specific case. I'm on it right now, but would love anybody that could help. 
![problem](https://github.com/user-attachments/assets/e8f00bb8-9bc5-4543-b623-d45c37647157)

abhigit-saha committed:
Fixed: Rendering issue of candidate words

abhigit-saha committed:
Merge remote-tracking branch 'origin/dev' into bip39wordlists

abhigit-saha committed:
Fixed indentation

abhigit-saha committed:
Merge branch 'SeedSigner:dev' into bip39wordlists

abhigit-saha commented:
The reason for me not putting a warning for SeedQR for other wordlists is because my initial plan is to have the wordlists as a way of encoding the english wordlist. The only reason I am holding onto that implementation is because there is still doubt amongst the developers regarding this. If anybody is reviewing this please go through https://medium.com/@abhijitsaha1006/the-problem-with-bitcoin-secret-words-f2b12c23ebe4. IMO it is something that we can incoporate but I need the approval of other devs as well (I got to know that this also was originally the idea of Jean Do after discussion in the telegram group)

abhigit-saha commented:
Also as per Keith's suggestion, I have tried shifting the location of the wordlists to the seedsigner-translations repo. Here is the gist regarding this (Not pushed to make the PR functional for now) 

https://gist.github.com/abhigit-saha/3a400ad13fc8b45e798aafbc48e9857b

kdmukai commented on src/seedsigner/views/settings_views.py:
```diff
@@ -98,7 +98,27 @@ def run(self):
+                title=_("Caution"),
+                status_headline=None,
+                text=_(
+                    "Non-English wordlists are not supported by a lot of wallets, "
+                    "and you could lose your bitcoin if you want to recover your seed "
+                    "with a wallet that does not support it."
```
```suggestion
                    "Non-English BIP-39 wordlists are not supported by most wallet software. "
                    "Your backup will not work in "
                    "wallets that don't support your language's wordlist."),
```

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -293,23 +293,22 @@ def map_network_to_embit(cls, network) -> str:
+    WORDLIST_LANGUAGE__ZH_HANS_CN = "zh_Hans_CN"
+    WORDLIST_LANGUAGE__ZH_HANT_TW = "zh_Hant_TW"
+
     ALL_WORDLIST_LANGUAGES = [
-        (WORDLIST_LANGUAGE__ENGLISH, "English"),
-        # (WORDLIST_LANGUAGE__CHINESE_SIMPLIFIED, "简体中文"),
-        # (WORDLIST_LANGUAGE__CHINESE_TRADITIONAL, "繁體中文"),
-        # (WORDLIST_LANGUAGE__FRENCH, "Français"),
-        # (WORDLIST_LANGUAGE__ITALIAN, "Italiano"),
-        # (WORDLIST_LANGUAGE__JAPANESE, "日本語"),
-        # (WORDLIST_LANGUAGE__KOREAN, "한국어"),
-        # (WORDLIST_LANGUAGE__PORTUGUESE, "Português"),
+        (WORDLIST_LANGUAGE__EN, "English"),
+        (WORDLIST_LANGUAGE__ES, "Español"),
+        (WORDLIST_LANGUAGE__FR, "Français"),
+        (WORDLIST_LANGUAGE__IT, "Italiano"),
+        (WORDLIST_LANGUAGE__PT, "Português"),
+        # (WORDLIST_LANGUAGE__ZH_HANS_CN, "简体中文"),
+        # (WORDLIST_LANGUAGE__ZH_HANT_TW, "繁體中文"),
```
It feels unnecessarily redundant to redefine locale constants that already exist. `LOCALE__SPANISH` etc seems sufficient as-is.

I would reuse our existing constants within the `ALL_WORDLIST_LANGUAGES` list.

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -699,6 +689,14 @@ class SettingsDefinition:
                       visibility=SettingsConstants.VISIBILITY__ADVANCED,
                       default_value=SettingsConstants.OPTION__ENABLED),
 
+        SettingsEntry(category=SettingsConstants.CATEGORY__SYSTEM,
+                      attr_name=SettingsConstants.SETTING__WORDLIST_LANGUAGE,
+                      abbreviated_name="wordlist_lang",
+                      display_name=_mft("Mnemonic language"),
+                      type=SettingsConstants.TYPE__SELECT_1,
+                      selection_options=SettingsConstants.ALL_WORDLIST_LANGUAGES,
+                      visibility=SettingsConstants.VISIBILITY__ADVANCED,
+                      default_value=SettingsConstants.WORDLIST_LANGUAGE__EN),
```
This should probably be relocated to be just after "BIP-39 passphrase" to keep the related options together.

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -699,6 +689,14 @@ class SettingsDefinition:
+        SettingsEntry(category=SettingsConstants.CATEGORY__SYSTEM,
+                      attr_name=SettingsConstants.SETTING__WORDLIST_LANGUAGE,
+                      abbreviated_name="wordlist_lang",
+                      display_name=_mft("Mnemonic language"),
```
Display name should match the naming convention elsewhere: "BIP-39 mnemonic language" or "BIP-39 wordlist" or maybe just "BIP-39 language"(?).

kdmukai commented:
Our documentation somewhere should point to the wordlist specification in the bip: https://github.com/bitcoin/bips/blob/master/bip-0039/bip-0039-wordlists.md

Along with a detailed discussion of the risks / tradeoffs / implementation details / etc.

kdmukai commented:
I don't know if it's been discussed earlier, but I would also be extremely concerned about including a wordlist in a language that is not defined in the BIP.

A Catalan wordlist may exist, but how do we know that it has reached consensus within the community or if any other tools have any plans to include it? Since even the wordlists in the BIP are barely supported (e.g. Spanish), it already feels like this feature is on shaky ground; expanding it to a language that is outside of the BIP's "schelling point" feels too dangerous.

Note: I realize that this PR does NOT include any languages that aren't included in the BIP. But I think we need clarity on how we intend to use this feature going forward if this PR is merged. Such decisions should also be included in the additional documentation I suggested above.

abhigit-saha commented:
> I don't know if it's been discussed earlier, but I would also be extremely concerned about including a wordlist in a language that is not defined in the BIP.
> 
> A Catalan wordlist may exist, but how do we know that it has reached consensus within the community or if any other tools have any plans to include it? Since even the wordlists in the BIP are barely supported (e.g. Spanish), it already feels like this feature is on shaky ground; expanding it to a language that is outside of the BIP's "schelling point" feels too dangerous.
> 
> Note: I realize that this PR does NOT include any languages that aren't included in the BIP. But I think we need clarity on how we intend to use this feature going forward if this PR is merged. Such decisions should also be included in the additional documentation I suggested above.

Hi Keith, thank you for taking the time to review this. As for the security considerations, having started that discussion in the telegram group, I had come up with this solution https://medium.com/@abhijitsaha1006/the-problem-with-bitcoin-secret-words-f2b12c23ebe4. This in essence means that we aren't strictly sticking to the bip39 spec, but a more generous version of it, that allows other language wordlists to be backed up by english wordlists. So even if another wallet doesn't support non-english wordlists, the user can still back up the words if they map their corresponding language's seed phrase words to the english ones.  Do let me know if I can go ahead with modifying that in the current pr

abhigit-saha committed:
Changed warning screen message

Co-authored-by: kdmukai <keith.mukai@essaytagger.com>

kdmukai commented:
> Do let me know if I can go ahead with modifying that in the current pr

That would be a GIANT feature change that would be VERY difficult to get any consensus on anytime soon.

It should probably be proposed as a BIP in order to try to get broader community buy-in. But even if it's received well, the BIP process is even slower than we are. Also bitcoin core considers bip39 "Universally not recommended for implementation" so I'm not sure how people would react to a follow-up BIP related to it.

That being said, if you do want to push for this, having a usable demo fork in SeedSigner might help you build more interest and momentum.

I'd suggest keeping this PR focused on the existing bip39 spec and getting it as polished up as possible for the Summer of Bitcoin deadline.  But also leverage it to create an alternate branch that implements your idea.

kdmukai commented:
I think I'm leaning toward two options:
* As mentioned above, this PR to add the bip39 alternate wordlists.
* And then a kind of compromise approach that is related to your bigger idea, but less ambitious: Just focus on creating backups by index number for an English mnemonic. The idea is that non-English speakers would never have to work with the English mnemonic itself (even though it remains an English mnemonic under the hood). Index numbers are an improvement over English bip39 backups for non-English speakers. Index numbers can be expressed in the local numbering system (e.g. `1147` == `١١٤٧`). But this less ambitious version avoids the complication of defining a new wordlist for each language. Different locales COULD define their own wordlist and fully implement your concept, but that can live outside of the SeedSigner world; when it comes time to input them, they only enter indices. I think words still offer a UX win over asking people to record numbers, but having a defined, agreed upon wordlist in each language just feels too daunting.

abhigit-saha committed:
Shifted wordlists files to translations repo

abhigit-saha committed:
Merge branch 'bip39wordlists' of https://github.com/abhigit-saha/seedsigner

abhigit-saha committed:
Replaced redundant wordlist constants with locale constants

abhigit-saha committed:
Shifted BIP-39 wordlist setting to below the passphrase

abhigit-saha committed:
Cleanup

abhigit-saha committed:
Merge with latest

abhigit-saha committed:
Removed old code

abhigit-saha committed:
Merge branch 'SeedSigner:dev' into bip39wordlists

abhigit-saha committed:
Renamed supported to wordlist and transferred constants to a separate file

abhigit-saha committed:
Added Error screen for non-english wordlists in ScanView

Chaitanya-Keyal commented on src/seedsigner/models/seed.py:
```diff
@@ -160,7 +156,6 @@ def get_bip85_child_mnemonic(self, bip85_index: int, bip85_num_words: int, netwo
         root = bip32.HDKey.from_seed(self.seed_bytes, version=NETWORKS[SettingsConstants.map_network_to_embit(network)]["xprv"])
 
-        # TODO: Support other BIP-39 wordlist languages!
         return bip85.derive_mnemonic(root, bip85_num_words, bip85_index)
```
I haven't followed the previous discussions here, but just looking at this part; I believe the `derive_mnemonic` function needs to accept a language code for `embit` to generate child mnemonics in different languages. Right now it defaults to English (0'), which could cause confusion if not adjusted. We likely just need to map the correct language codes based on the selected wordlist.

For reference, the [BIP-85 spec](https://github.com/bitcoin/bips/blob/master/bip-0085.mediawiki#bip39) provides additional details.

It may also make sense to keep the existing TODO comment for now and address this BIP-85 related change in a separate PR.

abhigit-saha replied:
I think this would be a PR suited more for embit than here. Since many embit functions would then require the language code, It would be a massive PR that is probably not mergeable since embit is so widely used and Bip39 (esp other language wordlists) so controversial.

Maybe we disable bip85 child seeds for non English ones (like for seedqr).

Tbh I do believe the BIP39 is very flawed and a mapping that encodes the inherent entropy to english ones (for non english mnemonic) is a better more intuitive and safer alternative that probably would only requirw a seedsigner module without having to change embit in any way.

I'll try to work on it soon though!

Chaitanya-Keyal replied:
`embit` already supports adding the language code for deriving child mnemonics: https://github.com/diybitcoinhardware/embit/blob/9d597749d6dfae19375836cf92d649012fc0298c/src/embit/bip85.py#L28-L35

> Maybe we disable bip85 child seeds for non English ones (like for seedqr).

However, this also sounds acceptable to me!

notTanveer commented:
got this error while testing this pr, for all the available languages in the BIP-39 wordlist setting, 

<img width="239" height="244" alt="image" src="https://github.com/user-attachments/assets/df6532f4-47c0-4374-af3e-75e83bc752ef" />

<br></br>

also consider adding bottom padding or write a short instruction note here,

<img width="239" height="244" alt="image" src="https://github.com/user-attachments/assets/92011dca-0e44-464c-a672-43d342912776" />
